### PR TITLE
OpenTelemetry HTTP & gRPC Listeners for Distributed Tracing

### PR DESCRIPTION
**Pull request by oppegard**
### Description
This feature enables the proxy to receive OpenTelemetry traces (via protobufs) over gRPC or HTTP listeners.

- The listeners are enabled with the new config options `otlpGrpcListenerPorts` and `otlpHttpListenerPorts`. They are not enabled by default.
- While the listeners only support trace data now, work is underway to enable them to receive OpenTelemetry metric data (via protobufs). This is the reasoning for not using the package namespace `com.wavefront.agent.listeners.tracing`, since these listeners won't be restricted to just traces.
- It also contains some basic refactoring within `PushAgent.java`.

### Testing Notes

To test this in an end-to-end manner, we've been using the [opentelemetry java agent](https://github.com/open-telemetry/opentelemetry-java-instrumentation#getting-started) with the [spring petclinic sample app](https://github.com/spring-projects/spring-petclinic).

First, build this branch of the proxy and enable the gRPC listener with `otlpGrpcListenerPorts=4317`.

After building the petclinic repo and downloading [opentelemetry-javaagent.jar](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/download/v1.11.1/opentelemetry-javaagent.jar), run it with the following:

```sh
# Assumes proxy is listening at localhost:4317
java -javaagent:/tmp/opentelemetry-javaagent.jar \
  -Dotel.service.name=petclinic \
  -Dotel.resource.attributes=application=pet-app  \
  -jar target/spring-petclinic*.jar
```
The above will result in WF traces with tags of `application=pet-app` and `service=petclinic`. At startup it will generate some traces related to in-memory DB migrations. After, you can visit localhost:8080 to generate traces demonstrating HTTP request/response work.

If testing the HTTP listener is desired, configure the proxy to run with `otlpHttpListenerPorts=4318` and run the petclinic app in the same way, but add the additional parameter of `-Dotel.exporter.otlp.protocol=http/protobuf`.

If verbose proxy logging is desired to view the ingested OTel structures and the resulting transformation into WF spans and span logs, the following config can be added to log4j:

```
    <Loggers>
        <Root level="INFO">
            <AppenderRef ref="Console" />
        </Root>
        <AsyncLogger name="OTLPDataLogger" level="ALL" additivity="false">
            <AppenderRef ref="Console"/>
       </AsyncLogger>
    </Loggers>
```

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -279,6 +279,13 @@
         <artifactId>jchronic</artifactId>
         <version>0.2.6</version>
       </dependency>
+      <dependency>
+        <groupId>io.opentelemetry</groupId>
+        <artifactId>opentelemetry-bom-alpha</artifactId>
+        <version>1.6.0-alpha</version>
```
Comment (laullon) [inline pom.xml:285]: Alpha ?

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java [hunk 2]
```diff
@@ -94,15 +96,6 @@
...
 
-import io.grpc.netty.NettyServerBuilder;
-import io.netty.channel.ChannelHandler;
-import io.netty.channel.ChannelOption;
-import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
-import io.netty.handler.codec.bytes.ByteArrayDecoder;
-import io.netty.handler.codec.http.HttpMethod;
-import io.netty.handler.codec.http.cors.CorsConfig;
-import io.netty.handler.codec.http.cors.CorsConfigBuilder;
-import io.netty.handler.ssl.SslContext;
```
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/PushAgent.java:105]: Can you revert this "format changes" to keep the diff smaller ?

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java [hunk 3]
```diff
@@ -588,10 +632,10 @@ protected void startPickleListener(String strPort,
         preprocessors.get(strPort), blockedPointsLogger);
 
     startAsManagedThread(port, new TcpIngester(createInitializer(ImmutableList.of(
-        () -> new LengthFieldBasedFrameDecoder(ByteOrder.BIG_ENDIAN, 1000000, 0, 4, 0, 4, false),
-        ByteArrayDecoder::new, () -> channelHandler), port,
+            () -> new LengthFieldBasedFrameDecoder(ByteOrder.BIG_ENDIAN, 1000000, 0, 4, 0, 4, false),
+            ByteArrayDecoder::new, () -> channelHandler), port,
```
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/PushAgent.java:636]: format

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java (hunk 2 above)
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/PushAgent.java]: Reverted.

File: pom.xml (hunk 1 above)
Comment (oppegard) [reply pom.xml:285]: The OTel community is stating this is alpha, even though the parts we're using (Tracing) is considered stable.

**Review by laullon — CHANGES_REQUESTED** [review]

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpHttpHandler.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    this.scheduledExecutorService =
+        Executors.newScheduledThreadPool(1, new NamedThreadFactory("otlp-http-heart-beater"));
+    scheduledExecutorService.scheduleAtFixedRate(this, 1, 1, TimeUnit.MINUTES);
+
+    this.internalReporter = OtlpProtobufUtils.createAndStartInternalReporter(sender);
+  }
+
+  @Override
+  protected void handleHttpMessage(ChannelHandlerContext ctx, FullHttpRequest request) throws URISyntaxException {
+//  TODO:  if request.path == "/v1/traces"
```
[the remark this answers is not in the document]
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpHttpHandler.java]: Resolved with https://github.com/wavefrontHQ/wavefront-proxy/pull/656/commits/847ad1ed89affffb51b4f149a9c016040e3ec7e6.

**Review by laullon — APPROVED** [review]

**Comment by ajayj89** [comment]
Team performed following 2 items:
- Performance tests over the weekend on docker proxy to make sure legacy features (points ingestion, etc) are working fine. @sbhakta-vmware 
- Validating the feature based on testing steps above. @anwang-vmware

**Review by ajayj89 — APPROVED** [review]

**Pull request merged**